### PR TITLE
MemoryLoad as Statement

### PR DESCRIPTION
**Pull request by l-kent**
This makes MemoryLoad into a Statement instead of an Expr - it now represents an assignment to a variable with the RHS being a load from memory. This is done by creating a new temporary variable to represent the loaded value, which becomes a separate LocalAssign.

This concept was originally proposed by @ailrst, and it is clear to me now that this will significantly simplify things for the analyses and make bugs easier to find, despite unfortunately making the parsers much more unwieldy. The main motivation behind this is to simplify things for adding region analysis to the DSA, as that requires extending it slightly to give results for what the index of each load points to, which is slightly distinct from finding the value of the LHS of an assignment with a MemoryLoad contained in in the assignment's RHS somewhere.

I have also done a fair amount of general clean up including:
-Fixing a deprecation warning in VariableDependencyAnalysis
-Consolidating the common parts of two separate constant propagation analyses into one trait extended by both
-Assign is renamed to LocalAssign (to allow Assign to be a trait including both MemoryLoad and Assign)
-MemoryAssign is renamed to MemoryStore

### CODE REVIEW

File: src/main/scala/analysis/GlobalRegionAnalysis.scala [hunk 1]
```diff
@@ -169,16 +168,13 @@ trait GlobalRegionAnalysis(val program: Program,
...
-          // this is a constant but we need to check if it is a data region
-          checkIfDefined(evalMemLoadToGlobal(assign.rhs, 1, assign), n)
-        }
+      case store: MemoryStore =>
+        checkIfDefined(evalMemLoadToGlobal(store.index, store.size, store), n)
+      case load: MemoryLoad =>
+        checkIfDefined(evalMemLoadToGlobal(load.index, load.size, load, loadOp = true), n)
+      case assign: LocalAssign =>
+        // this is a constant but we need to check if it is a data region
+        checkIfDefined(evalMemLoadToGlobal(assign.rhs, 1, assign), n)
```
Comment (ailrst) [inline src/main/scala/analysis/GlobalRegionAnalysis.scala:177]: `unwrapExpr` has disappeared

**Review by ailrst — CHANGES_REQUESTED** [review]

File: src/main/scala/ir/Statement.scala [hunk 2]
```diff
@@ -23,33 +23,47 @@ sealed trait Statement extends Command, IntrusiveListElement[Statement] {
...
   def successor: Command = parent.statements.nextOption(this).getOrElse(parent.jump)
+}
 
+sealed trait Assign extends Statement {
+  var lhs: Variable
 }
 
-// invariant: rhs contains at most one MemoryLoad
-class Assign(var lhs: Variable, var rhs: Expr, override val label: Option[String] = None) extends Statement {
+class LocalAssign(var lhs: Variable, var rhs: Expr, override val label: Option[String] = None) extends Assign {
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/ir/Statement.scala:33]: It's necessary for the ReachingDefinitionsAnalysis.

File: src/main/scala/analysis/GlobalRegionAnalysis.scala [hunk 3]
```diff
@@ -88,7 +88,6 @@ trait GlobalRegionAnalysis(val program: Program,
           } else {
             Set()
           }
-        case _: MemoryLoad => ???
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/analysis/GlobalRegionAnalysis.scala:91]: MemoryLoad is no longer an Expr

File: src/main/scala/analysis/TaintAnalysis.scala [hunk 4]
```diff
@@ -69,25 +72,23 @@ def getMemoryVariable(
...
-    case BinaryExpr(BVADD, arg1, arg2) if arg1 == stackPointer => {
+    case BinaryExpr(BVADD, arg1, arg2) if arg1 == stackPointer =>
       evaluateExpression(arg2, constProp(n)) match
         // TODO This assumes that all stack variables are initialized local variables, which is not necessarily the case.
         //      If a stack address is read, without being assigned a value in this procedure, it will be
         //      assumed untainted, when in reality it may be UnknownMemory.
         //case Some(addr) => Some(LocalStackVariable(addr, size))
-        case Some(addr) => None
+        case Some(_) => None
         case None => None
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/analysis/TaintAnalysis.scala:82]: That's what it already was doing, seems to be a placeholder currently since there's the commented out version.

File: src/main/scala/analysis/GlobalRegionAnalysis.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ailrst) [reply src/main/scala/analysis/GlobalRegionAnalysis.scala:91]: Of course

File: src/main/scala/analysis/TaintAnalysis.scala [hunk 5]
```diff
@@ -69,25 +72,23 @@ def getMemoryVariable(
 
   expression match {
     // TODO assumes stack var accesses are all of the form R31 + n, or just R31, when in reality they could be more complex.
-    case BinaryExpr(BVADD, arg1, arg2) if arg1 == stackPointer => {
+    case BinaryExpr(BVADD, arg1, arg2) if arg1 == stackPointer =>
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/analysis/TaintAnalysis.scala:75]: We were never using parentheses after `case =>`, that has never been standard, but elsewhere yes. Preferring parentheses over indented blocks is about keeping with Scala 2-style formatting, not the Python-style Scala 3 which only uses indentation. The Scala 2 style does not use parentheses for match cases either.

What's the advantage for refactoring here? The clear delineation of match cases means I don't find there to be an issue with the lack of parentheses unlike in other cases.

File: src/main/scala/translating/BAPToIR.scala [hunk 6]
```diff
@@ -77,17 +82,138 @@ class BAPToIR(var program: BAPProgram, mainAddress: BigInt) {
...
+        val assign = LocalAssign(lhs, rhs, Some(b.line + "$1"))
+        Seq(loadWithLabel, assign)
+      } else {
+        val assign = LocalAssign(lhs, rhs, Some(b.line))
+        Seq(assign)
+      }
   }
 
+  private def translateExpr(e: BAPExpr): (Expr, Option[MemoryLoad]) = e match {
+    case BAPConcat(left, right) => (BinaryExpr(BVCONCAT, translateExprOnly(left), translateExprOnly(right)), None)
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/translating/BAPToIR.scala:106]: It's a bit crude at the moment, yes, mostly to avoid having to consider the case where both sides of a BinaryExpr could contain a load (which the lifter should not produce), since I was unaware of any cases where a Concat could contain a load to begin with. This is easy enough to fix though.

File: src/main/scala/ir/Expr.scala [hunk 7]
```diff
@@ -302,11 +296,11 @@ case class MemoryLoad(mem: Memory, index: Expr, endian: Endian, size: Int) exten
   }
   override def variables: Set[Variable] = index.variables
   override def gammas: Set[Variable] = Set()
-  override def loads: Set[MemoryLoad] = Set(this)
   override def getType: IRType = BitVecType(size)
   override def toString: String = s"MemoryLoad($mem, $index, $endian, $size)"
   override def acceptVisit(visitor: Visitor): Expr = visitor.visitMemoryLoad(this)
 }
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/ir/Expr.scala:302]: What?

[the remark this answers is not in the document]
Comment (ailrst) [reply src/main/scala/ir/Expr.scala:302]: This is just commented out, we can delete it

File: src/main/scala/util/RunUtils.scala [hunk 8]
```diff
@@ -342,40 +342,40 @@ object StaticAnalysis {
...
-    val newCPResult: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]] = ilcpsolver.analyze()
+    val intraProcConstProp = IntraProcConstantPropagation(IRProgram)
+    val intraProcConstPropResult: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]] = intraProcConstProp.analyze()
 
-    config.analysisResultsPath.foreach(s =>
-      writeToFile(printAnalysisResults(IRProgram, newCPResult), s"${s}_new_ir_constprop$iteration.txt")
-    )
+    config.analysisResultsPath.foreach { s =>
+      writeToFile(printAnalysisResults(IRProgram, intraProcConstPropResult), s"${s}_new_ir_constprop$iteration.txt")
+    }
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/util/RunUtils.scala:363]: I am trying to avoid unnecessary parentheses that do not aid in the clarity of the code since there are multiple on one line.

File: src/main/scala/analysis/TaintAnalysis.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (ailrst) [reply src/main/scala/analysis/TaintAnalysis.scala:75]: With two spaces of indentation scope is practically unreadable with pure indentation, especially when we get dozens of nested match statements as we tend to (its own problem). Imo omitting the parenth is only acceptable if the block is one line. Given its just an aesthetic preference its not worth arguing over, just don't arbitrarily change it in existing code in unrelated PRs as it is just diff and merge noise.

[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/analysis/TaintAnalysis.scala:75]: For case clauses, they are clearly delineated by each `case` and the match statement still has the surrounding braces that clearly marks where the case clauses end, so I don't find there to be an issue for that case. When match statements don't have braces then it's hard to follow but the case clauses are fine without them.

The meaningful indentation that should be avoided and plays very badly with IDEs is the syntax that was introduced in Scala 3: https://docs.scala-lang.org/scala3/reference/other-new-features/indentation.html

**Review by ailrst — APPROVED** [review]
